clippy_lints/src/uninit_vec.rs

```diff
@@ -0,0 +1,174 @@
+        for w in block.stmts.windows(2) {
+            if let StmtKind::Expr(expr) | StmtKind::Semi(expr) = w[1].kind {
+                handle_pair(cx, &w[0], expr);
+            }
+        }
+
+        if let (Some(stmt), Some(expr)) = (block.stmts.last(), block.expr) {
+            handle_pair(cx, stmt, expr);
+        }
+    }
+}
+
+fn handle_pair(cx: &LateContext<'tcx>, first: &'tcx Stmt<'tcx>, second: &'tcx Expr<'tcx>) {
+    if_chain! {
+        if let Some(vec) = extract_with_capacity_or_reserve_target(cx, first);
+        if let Some((set_len_self, call_span)) = extract_set_len_self(cx, second);
+        if vec.eq_expr(cx, set_len_self);
+        if let ty::Ref(_, vec_ty, _) = cx.typeck_results().expr_ty_adjusted(set_len_self).kind();
+        if let ty::Adt(_, substs) = vec_ty.kind();
+        // Check T of Vec<T>
+        if !is_uninit_ty_valid(cx, substs.type_at(0));
+        then {
+            span_lint_and_note(
+                cx,
+                UNINIT_VEC,
+                call_span,
+                "calling `set_len()` immediately after reserving a buffer creates uninitialized values",
+                Some(first.span),
+                "the buffer is reserved here"
+            );
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum LocalOrExpr<'tcx> {
+    Local(HirId),
+    Expr(&'tcx Expr<'tcx>),
+}
+
+impl<'tcx> LocalOrExpr<'tcx> {
+    fn eq_expr(self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) -> bool {
+        match self {
+            LocalOrExpr::Local(hir_id) => path_to_local_id(expr, hir_id),
+            LocalOrExpr::Expr(self_expr) => SpanlessEq::new(cx).eq_expr(self_expr, expr),
+        }
+    }
+}
+
+/// Returns the target vec of `Vec::with_capacity()` or `Vec::reserve()`
+fn extract_with_capacity_or_reserve_target(cx: &LateContext<'_>, stmt: &'tcx Stmt<'_>) -> Option<LocalOrExpr<'tcx>> {
+    match stmt.kind {
+        StmtKind::Local(local) => {
+            // let mut x = Vec::with_capacity()
+            if_chain! {
+                if let Some(init_expr) = local.init;
+                if let PatKind::Binding(_, hir_id, _, None) = local.pat.kind;
+                if is_with_capacity(cx, init_expr);
+                then {
+                    Some(LocalOrExpr::Local(hir_id))
+                } else {
+                    None
+                }
+            }
+        },
+        StmtKind::Expr(expr) | StmtKind::Semi(expr) => {
+            match expr.kind {
+                ExprKind::Assign(lhs, rhs, _span) if is_with_capacity(cx, rhs) => {
+                    // self.vec = Vec::with_capacity()
+                    Some(LocalOrExpr::Expr(lhs))
+                },
+                ExprKind::MethodCall(_, _, [vec_expr, _], _) => {
+                    // self.vec.reserve()
+                    if_chain! {
+                        if let Some(id) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
+                        if match_def_path(cx, id, &paths::VEC_RESERVE);
```

Comment:
We try to move away from paths to check for method calls. Here it would be better to first check for the `self` type and the check the name of the method. This could look like this:

```suggestion
                ExprKind::MethodCall(path, _, [vec_expr, _], _) => {
                    // self.vec.reserve()
                    if_chain! {
                        if is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(self_expr).peel_refs() , sym::vec_type);
                        if path.ident.name.as_str() == "reserve";
```

And I would probably not use an `if_chain!` here to reduce indention :upside_down_face: